src/liballoc/heap.rs

```diff
@@ -44,6 +45,62 @@ fn check_size_and_alignment(size: usize, align: usize) {
                   align);
 }
 
+#[derive(Copy, Clone, Default, Debug)]
+pub struct HeapAllocator;
+
+unsafe impl Allocator for HeapAllocator {
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
+        let addr = allocate(layout.size(), layout.align());
+        if addr.is_null() {
+            Err(AllocErr::Exhausted { request: layout })
+        } else {
+            Ok(addr)
+        }
+    }
+
+    unsafe fn dealloc(&mut self, ptr: *mut u8, layout: Layout) {
+        deallocate(ptr, layout.size(), layout.align());
+    }
+
+    unsafe fn usable_size(&self, layout: &Layout) -> (usize, usize) {
+        (layout.size(), usable_size(layout.size(), layout.align()))
+    }
+
+    unsafe fn realloc(&mut self,
+                      ptr: *mut u8,
+                      layout: Layout,
+                      new_layout: Layout)
+                      -> Result<*mut u8, AllocErr>
+    {
+        if layout.align() != new_layout.align() {
+            return Err(AllocErr::Unsupported { details: "mismatched input layouts" });
+        }
+        let addr = reallocate(ptr, layout.size(), new_layout.size(), layout.align());
+        if addr.is_null() {
+            Err(AllocErr::Exhausted { request: new_layout })
+        } else {
+            Ok(addr)
+        }
+    }
+
+    unsafe fn grow_in_place(&mut self,
+                            ptr: *mut u8,
+                            layout: Layout,
+                            new_layout: Layout)
+                            -> Result<(), CannotReallocInPlace>
+    {
+        // grow_in_place spec requires this, and the spec for reallocate_inplace
+        // makes it hard to detect failure if it does not hold.
+        debug_assert!(new_layout.size() >= layout.size());
+
+        if layout.align() != new_layout.align() { // reallocate_inplace requires this.
```

Comment:
... but that's exactly the issue. `reallocate_inplace` doesn't take enough inputs to deal with distinct old and new alignments. It just says that its required to take the old alignment as its `align` parameter, and that it will produce a block that is aligned according to that.

So if you give this method distinct alignments for old and new, then `reallocate_inplace` is not going to be able to satisfy the request.